enterprise/server/backends/distributed/distributed.go

```diff
@@ -390,13 +390,25 @@ func dedupe(in []string) []string {
 // peers are returned in random order.
 func (c *Cache) readPeers(d *repb.Digest) *peerset.PeerSet {
 	peers := c.consistentHash.GetAllReplicas(d.GetHash())
-	primaryPeers := peers[:c.config.ReplicationFactor]
-	secondaryPeers := peers[c.config.ReplicationFactor:]
+	var primaryPeers, secondaryPeers []string
+	if len(peers) < c.config.ReplicationFactor {
+		primaryPeers = []string{}
+		secondaryPeers = []string{}
+	} else {
+		primaryPeers = peers[:c.config.ReplicationFactor]
+		secondaryPeers = peers[c.config.ReplicationFactor:]
+	}
 
 	if len(c.config.ExtraNodes) > 0 {
 		extendedPeerList := c.extraConsistentHash.GetAllReplicas(d.GetHash())
-		allPrimaryPeers := extendedPeerList[:c.config.ReplicationFactor]
-		allSecondaryPeers := extendedPeerList[c.config.ReplicationFactor:]
+		var allPrimaryPeers, allSecondaryPeers []string
+		if len(extendedPeerList) < c.config.ReplicationFactor {
+			allPrimaryPeers = []string{}
+			allSecondaryPeers = []string{}
+		} else {
+			allPrimaryPeers = extendedPeerList[:c.config.ReplicationFactor]
+			allSecondaryPeers = extendedPeerList[c.config.ReplicationFactor:]
+		}
```

Comment:
I think it would make more sense to move this check (that replication factor <= peer count) to the health check -- that way the cache emits a warning and does not become healthy if misconfigured.

(that's the `func (c *Cache) Check(ctx context.Context) error {` method)

This check should happen even if the nodelist is hardcoded btw.